### PR TITLE
Warn about variable number of dependencies

### PR DESCRIPTION
**Pull request by acdlite**
We don't check this in prod, since best practice is to always pass
these inline. But we should still warn in dev.

I also moved all the existing Hook tests to a separate file called ReactHooksWithNoopRenderer. New tests should use test renderer.

Second commit is the relevant one.

### CODE REVIEW

File: packages/react-reconciler/src/ReactFiberHooks.js [hunk 1]
```diff
@@ -760,8 +761,19 @@ function dispatchAction<S, A>(
 }
 
 function inputsAreEqual(arr1, arr2) {
-  // Don't bother comparing lengths because these arrays are always
+  // Don't bother comparing lengths in prod because these arrays should be
   // passed inline.
+  if (__DEV__) {
+    warning(
+      arr1.length === arr2.length,
+      'Detected a variable number of dependencies. The length of the ' +
```
Comment (bvaughn) [inline packages/react-reconciler/src/ReactFiberHooks.js]: Seems like this error message should mention "hooks". It's a little cryptic.

Comment (acdlite) [reply packages/react-reconciler/src/ReactFiberHooks.js]: Perhaps "variable number of dependencies to a hook"?

Comment (acdlite) [reply packages/react-reconciler/src/ReactFiberHooks.js]: Updated

Comment (bvaughn) [reply packages/react-reconciler/src/ReactFiberHooks.js]: Yeah, I like that change.

**Comment by acdlite** [comment]
Going to merge without squashing to preserve the history

**Pull request merged**